contents/security/sqlInjection.md

```diff
@@ -0,0 +1,216 @@
+
+
+## Methods of SQL Injection ##
+
+There are two common methods for attackers to identify and exploit an SQL injection weakness
+in websites:
+
+### SQL Injection ###
+
+The attacker makes use of errors generated from executed queries. The errors from the 
+database allows the attacker to fine-tune their attack so their next attempt would 
+be more successful.
+
+For example, if there is a shopping website which displays products from a database, 
+it may generate a query similar to this:
+
+```html
+String query = "SELECT * FROM Products WHERE productName = " + input_searchTerm; 
+```
+
+The attacker may try to use this vulnerability to retrieve users from the database. A 
+search term such as "1 UNION SELECT * FROM Users" may be used. This would result in the 
+following query being executed:
+
+```html
+SELECT * FROM Products WHERE productName = 1 UNION
+SELECT * FROM Users
+```
+
+Union is an operator in SQL, which allows a query to combine multiple queries into one. 
+This attack attempts to use the products display on the website to display user data instead. 
+If the attacker receives the following message:
+
+```html
+ORA-00942: table or view does not exist.
+```
+
+The attacker will know that there is no table named "Users" in the database. If, after a number 
+of tries, a different message appears:
+
+```html
+ORA-01789: query block has incorrect number of result columns.
+```
+
+Then the attacker would have found a table which exists in the database, as the error message 
+has changed. Now that an existing table has been found, the attacker can refine the query further. 
+In this example, the attack failed because the number of columns was inconsistent with the expected 
+result. The attacker may try "SELECT Username, Password" next, which will limit the retrieved data 
+to be two columns, one for each parameter requested.
+
+
+### Blind SQL Injection ###
+
+Even when there is no detailed error from the database displayed, attackers can gather 
+information though other means, such as by testing different true or false statements.
+
+When attempting to login to a website normally, the attacker may see this error message:
+
+```html
+Error: Invalid credentials!
+```
+
+The message does not give enough information to the attacker. It is unknown whether 
+the username or password is wrong here. However, this is not enough to deter a SQL 
+injection attack! If the following attack is attempted and successful:
+
+```html
+Username: foo
+Password: bar’ OR ‘’=’
+```
+
+Then the attacker knows that the form is vulnerable to following attacks. Once a SQL 
+injection vulnerability is discovered, the attacker can continue to exploit the 
+database as per normal SQL Injection. For example, if the attacker knew of a table 
+in the database named 'Users', the attacker can enter this malicious string, which 
+would delete the entire table and its associated data:
+
+```html
+Username: foo
+Password: bar’; DROP TABLE Users;--
```

Comment:
I don't think this is an example of a blind sql injection. The purpose of blind sql injection attack is to steal data, not to delete data. The point of blind sql is that the error message is generic so the attacker cannot see the actual data he wants to see